source/common/router/router.cc

```diff
@@ -1446,10 +1481,21 @@ void Filter::doRetry() {
   pending_retries_--;
   UpstreamRequestPtr upstream_request;
 
-  Http::ConnectionPool::Instance* conn_pool = getHttpConnPool();
-  if (conn_pool) {
-    upstream_request =
-        std::make_unique<UpstreamRequest>(*this, std::make_unique<HttpConnPool>(*conn_pool));
+  if (!shouldTcpProxy(*downstream_headers_, route_entry_)) {
```

Comment:
This if/else block is mostly duplicated from decodeHeaders and is a little tricky especially with the transport socket options stuff so it might be worth it to share, but up to you.